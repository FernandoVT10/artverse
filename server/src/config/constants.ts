import { config } from "dotenv";
// set all variables from ".env" file in the "process.env"
config();

type DBConfig = {
  host: string;
  username: string;
  password: string;
  databaseName: string;
  port: number;
};

const dbPort = parseInt(process.env.PORT || "") || 3306;

export const DB_CONFIG: DBConfig = {
  host: process.env.DB_HOST || "",
  username: process.env.DB_USERNAME || "",
  password: process.env.DB_PASSWORD || "",
  databaseName: process.env.DB_NAME || "",
  port: dbPort,
};
