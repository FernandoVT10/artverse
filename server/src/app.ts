import express from "express";
import bodyParser from "body-parser";
import routes from "./routes";
import errorHandler from "@middlewares/errorHandler";

import { PUBLIC_DIR } from "./config/constants";

const app = express();

app.use(express.static(PUBLIC_DIR));
app.use(bodyParser.json());

app.use(routes);
app.use(errorHandler());

export default app;
