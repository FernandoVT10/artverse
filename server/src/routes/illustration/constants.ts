import path from "path";
import { IMAGES_DIR } from "@config/constants";
import { Size } from "@utils/images/saveAndResizeMulterFileAsImage";

export const ILLUSTRATIONS_DESTINATION = path.resolve(
  IMAGES_DIR,
  "./illustrations/"
);

export const ILLUSTRATIONS_SIZES: Size[] = [
  {
    width: 250,
    height: 250,
    suffix: "thumbnail",
  },
];
