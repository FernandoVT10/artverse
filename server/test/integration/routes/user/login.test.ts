import jwt from "jsonwebtoken";
import bcrypt from "bcrypt";
import supertest from "supertest";
import app from "@app";

import { testPasswordField } from "./shared";
import { clearDB } from "@test-utils/setupDB";
import { createUser } from "@test-utils/factories/userFactory";
import { JWT_SECRET_KEY } from "@config/constants";
import { User } from "@models";

describe("Integration POST /api/users/login", () => {
  let user: User;

  const usernameOrEmail = "foo-bar";
  const plainPassword = "secret";

  beforeEach(async () => {
    await clearDB();

    const hashedPassword = await bcrypt.hash(plainPassword, 1);
    user = await createUser({
      username: usernameOrEmail,
      password: hashedPassword,
    });
  });

  const request = supertest(app);
  const requestAPI = () => request.post("/api/users/login");

  const getTokenFromCookies = (cookies: string[]): string => {
    const cookie = cookies[0];
    const token = cookie.split(";")[0];
    const parsedToken = token.replace("token=", "");
    return parsedToken;
  };

  it("should set cookie with a valid jsonwebtoken and return a success message", async () => {
    const res = await requestAPI()
      .send({
        usernameOrEmail: user.username,
        password: plainPassword,
      })
      .expect(200);

    expect(res.body.success).toBeTruthy();

    const token = getTokenFromCookies(res.get("Set-Cookie"));

    expect(() => jwt.verify(token, JWT_SECRET_KEY)).not.toThrow();
  });

  it("should return an error when the password is incorrect", async () => {
    const res = await requestAPI()
      .send({
        usernameOrEmail: user.username,
        password: "123",
      })
      .expect(400);

    const { errors } = res.body;
    expect(errors).toMatchSnapshot();
  });

  describe("validation", () => {
    describe("usernameOrEmail field", () => {
      it("should return an error when the username or email don't exist", async () => {
        const res = await requestAPI().send({
          usernameOrEmail: "foo",
          password: "test",
        });

        const { errors } = res.body;

        expect(errors).toMatchSnapshot();
      });
    });

    testPasswordField(requestAPI, {
      // here i'm passing an username that already exists to not have
      // an error with the usernameOrEmail field
      usernameOrEmail,
    });
  });
});
