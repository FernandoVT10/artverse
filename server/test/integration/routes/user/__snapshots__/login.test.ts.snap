// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration routes/user/login should return an error when the password is incorrect 1`] = `undefined`;

exports[`Integration routes/user/login validation password field should return an error when the password is empty 1`] = `
Array [
  Object {
    "field": "password",
    "message": "The password is required",
  },
]
`;

exports[`Integration routes/user/login validation password field should return an error when the password is not a string 1`] = `
Array [
  Object {
    "field": "password",
    "message": "The password must be a string",
  },
]
`;

exports[`Integration routes/user/login validation usernameOrEmail field should return an error when the username or email don't exist 1`] = `
Array [
  Object {
    "field": "usernameOrEmail",
    "message": "The username or email don't exist",
  },
]
`;
